feat(ui): add extension the top-bar action menu (#19620)
argoproj/argo-cd: docs/developer-guide/extensions/ui-extensions.md

Below is an example of an extension using the flyout widget:

> 

Reviewer: ```suggestion
```

unnecessary extra new line added (this would also fail `markdownlint`, which I guess isn't running?)

```javascript
((window) => {
  const component = (props: {
    openFlyout: () => any
  }) => {
    return React.createElement(

>             "div",
>             {
>               style: { padding: "10px" },
>               onClick: () => props.openFlyout()
>             },
>             "Hello World"

Reviewer: ```suggestion
      "div",
      { 
        style: { padding: "10px" },
        onClick: () => props.openFlyout()
      },
      "Hello World"
```

unnecessary indentation as well, and it's also incorrect and inconsistent -- this adds a 4 space indent in one place to what is otherwise a 2 space code block...

    );
  };
  const flyout = () => {
    return React.createElement(
            "div",
            { style: { padding: "10px" } },

>             "This is a flyout"

Reviewer: same as above

    );
  };
  window.extensionsAPI.registerStatusPanelExtension(
          component,
          "My Extension",
          "my_extension",

>           flyout

Reviewer: same as above

When the extension button is clicked, the custom widget will be rendered in a flyout panel.

The extension should be registered using the `extensionsAPI.registerTopBarActionMenuExt` method:

```typescript
registerTopBarActionMenuExt(
  component: TopBarActionMenuExtComponent,
  title: string,
  id: string,
  flyout?: ExtensionComponent,
  shouldDisplay: (app?: Application) => boolean = () => true,
  iconClassName?: string,
  isMiddle = false
)
```

The callback function `shouldDisplay` should return true if the extension should be displayed and false otherwise:

```typescript
const shouldDisplay = (app: Application) => {
  return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
};
```

Below is an example of a simple extension with a flyout widget:

```javascript
((window) => {
  const shouldDisplay = () => {
    return true;
  };
  const flyout = () => {
    return React.createElement(
            "div",
            { style: { padding: "10px" } },

>             "This is a flyout"

Reviewer: same as above

    );
  };
  const component = () => {
    return React.createElement(
            "div",
            {
              onClick: () => flyout()
            },

>             "Toolbar Extension Test"

Reviewer: same as above

    );
  };
  window.extensionsAPI.registerTopBarActionMenuExt(
          component,
          "Toolbar Extension Test",
          "Toolbar_Extension_Test",
          flyout,
          shouldDisplay,
          '',

>           true

Reviewer: same as above. this also undoes a previous suggestion I made that you added...

Below is an example of an extension using the flyout widget:

> ```typescript

Reviewer: This is also an example, so my suggestion was to make this annotation JS and remove the type annotations on the props. I wouldn't change any other formatting or ordering and keep the diff light (right now you split out the code blocks etc, which I wouldn't recommend and actually feels confusing to me)

  );
})(window);
```

## Top Bar Action Menu Extensions

The top bar panel is the action menu at the top of the application view where the action buttons are displayed like Details, Sync, Refresh. Argo CD allows you to add new button to the top bar action menu of an application.
When the extension button is clicked, the custom widget will be rendered in a flyout panel.

The extension should be registered using the `extensionsAPI.registerTopBarActionMenuExt` method:

```typescript
registerTopBarActionMenuExt(
        component: TopBarActionMenuExtComponent,
        title: string,
        id: string,
        flyout?: ExtensionComponent,
        shouldDisplay: (app?: Application) => boolean = () => true,
        iconClassName?: string,
        isMiddle = false
)
```

The callback function `shouldDisplay` should return true if the extension should be displayed and false otherwise:

```typescript
const shouldDisplay =  (app: Application) => {
  return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
```

> Below is an example of a simple extension with a flyout widget:

Reviewer: ```suggestion

Below is an example of a simple extension with a flyout widget:
```

missed the new line in my last comment

  );
})(window);
```

## Top Bar Action Menu Extensions

The top bar panel is the action menu at the top of the application view where the action buttons are displayed like Details, Sync, Refresh. Argo CD allows you to add new button to the top bar action menu of an application.
When the extension button is clicked, the custom widget will be rendered in a flyout panel.

The extension should be registered using the `extensionsAPI.registerTopBarActionMenuExt` method:

```typescript
registerTopBarActionMenuExt(
        component: TopBarActionMenuExtComponent,
        title: string,
        id: string,
        flyout?: ExtensionComponent,
        shouldDisplay: (app?: Application) => boolean = () => true,
        iconClassName?: string,
        isMiddle = false
)
```

> Developer need to pass the function `shouldDisplay` which will be called to determine if the extension should be displayed or not. 
> The function should return true if the extension should be displayed and false otherwise.

Reviewer: ```suggestion

The callback function `shouldDisplay` should return true if the extension should be displayed and false otherwise:
```

- add a new line after the code block, per markdownlint
- simplify the text here which had several redundant phrases, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-simple-and-direct-language)
- add a colon to segue into the next code block

```typescript
const shouldDisplay =  (app: Application) => {
  return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
```

Below is an example of a simple extension with a flyout widget:

> ```typescript

Reviewer 2: Since the extensions need to be in javascript, so I think that in this doc, we want users to quickly get these samples up and running. And 'out of the box' (in this case, 'out of the docs'), they should just work.  The Application Status Panel extension sample WITHOUT flyout works fine as is.  I can copy the contents and put them in the extension*.js file in the /tmp/extensions folder, and it works nicely.  However, with the flyout, it doesn't work. This example doesn't work either (registerTopBarExtension is the old name?)

So, I think we could just provide the JS for this example. I got the following to work and there were no errors when the flyout appeared.  Feel free to update the md file with this, and rename the titles/names/etc. if you want.

```
((window) => {
    const shouldDisplay = () => {
      return true;
    };
    const flyout = () => {
      return React.createElement(
              "div",
              { style: { padding: "10px" } },
              "This is a flyout"
      );
    };
    const component = () => {
        return React.createElement(
                "div",
                {
                  onClick: () => flyout()
                },
                "Toolbar Extension Test"
        );
      };
    window.extensionsAPI.registerTopBarActionMenuExt(
            component,
            "Toolbar Extension Test",
            "Toolbar_Extension_Test",
            flyout,
            shouldDisplay,
            '',
            true
    );
  })(window);
```

Also, I got your https://github.com/ashutosh16/argocd-action-menu-extension/tree/main example to work, but I needed to tweak some things there. There's a CORS issue.

Reviewer 2: ![toolbarExtensionTest](https://github.com/user-attachments/assets/638fe064-2dba-4591-a9ed-626679013557)

Reviewer: > Since the extensions need to be in javascript, I think that in this doc, we want users to quickly get these samples up and running.

good point, nice catch! some of the equivalent code blocks are annotated as TS code blocks instead of JS above, so I think it's currently inconsistent in the docs. I'd probably suggest rewriting the "Flyout widget" block in JS. The "Application Status" and "System Level" code blocks are also annotated as TS, but are actually pure JS, so I'd suggest changing the syntax highlight there too for consistency.

In other words, only function type signatures on this page would be in TS, but actual example usage is in JS

Author: @agilgur5  updated the docs , pls validate

>   return application.metadata?.labels && application.metadata.labels['application.environmentLabelKey'] === "prd";

Reviewer: ```suggestion
  return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
```

could similarly simplify this with more optional chaining

> ``

Reviewer: ```suggestion

```

extraneous backticks

  );
})(window);
```

## Top Bar Action Menu Extensions

The top bar panel is the action menu at the top of the application view where the action buttons are displayed like Details, Sync, Refresh. Argo CD allows you to add new button to the top bar action menu of an application.
When the extension button is clicked, the custom widget will be rendered in a flyout panel.

The extension should be registered using the `extensionsAPI.registerTopBarActionMenuExt` method:

```typescript
registerTopBarActionMenuExt(

>         component: TopBarActionMenuExtComponent,
>         title: string,
>         id: string,
>         flyout?: ExtensionComponent,
>         shouldDisplay: (app?: Application) => boolean = () => true,
>         iconClassName?: string,
>         isMiddle = false

Reviewer: ```suggestion
  component: TopBarActionMenuExtComponent,
  title: string,
  id: string,
  flyout?: ExtensionComponent,
  shouldDisplay: (app?: Application) => boolean = () => true,
  iconClassName?: string,
  isMiddle = false
```

same as above, consistent 2 space indentation

)
```

The callback function `shouldDisplay` should return true if the extension should be displayed and false otherwise:

```typescript

>   const shouldDisplay =  (app: Application) => {
>     return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
>   };

Reviewer: ```suggestion
const shouldDisplay = (app: Application) => {
  return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
};
```

remove some unnecessary spaces

  const flyout = () => {
    return React.createElement(
            "div",
            { style: { padding: "10px" } },
            "This is a flyout"
    );
  };
  const component = () => {
    return React.createElement(
            "div",
            {
              onClick: () => flyout()
            },
            "Toolbar Extension Test"
    );

>   };

Reviewer: I'd move the `flyout` and `component` into the example below as they are examples and pure JS. I would also keep the formatting identical to the prior examples

whereas the `shouldDisplay` part above this can remain in TS as it shows a type signature, same as the registration function

```

Below is an example of a simple extension with a flyout widget:

```javascript
((window) => {
  const shouldDisplay = () => {
    return true;
  };
  window.extensionsAPI.registerTopBarActionMenuExt(

>           component,
>           "Toolbar Extension Test",
>           "Toolbar_Extension_Test",
>           flyout,
>           shouldDisplay,
>           '',
>           true

Reviewer: ```suggestion
    component,
    "Toolbar Extension Test",
    "Toolbar_Extension_Test",
    flyout,
    shouldDisplay,
    '',
    true
```

keep the indentation consistent with the code above as well as the other examples on this page: 2 space
